parl/framework/algorithm.py

```diff
@@ -0,0 +1,187 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle.fluid as fluid
+import parl.layers as layers
+import numpy as np
+
+
+class AlgorithmWrapper(object):
+    """
+    AlgorithmWrapper is responsible for I/O between the Algorithm and multiple Robots.
+    """
+    pass
+
+
+class Algorithm(object):
```

Comment:
Feel that all these should belong to AlgorithmWrapper. We should not expect that the user write so much stuff for an algorithm. The user should only need to take care of learn and predict for an algorithm.

Reply:
I was thinking that Algorithm is generally fixed. Then user only needs to implement the Model class. AW is responsible for data I/O.

Reply:
For different algorithms, according to the current implementation, the differences are only: 1) some hyperparameters of the optimizer (learning rate, learning algorithm, etc), and 2) the model object. Other than these two things, I believe all tasks can be represented by this general control flow of Algorithm so that the user can always use it without changes.